### ISSUE TITLE
Option to change behaviour of mouse back button

### ISSUE THREAD

**Issue by tomaszg7** [BestImageViewer/geeqie#659]
### ISSUE TYPE
* Patch

Hi,

I grew unsatisfied with my patch in issue https://github.com/BestImageViewer/geeqie/issues/512 (http://geeqie.org/cgi-bin/gitweb.cgi?p=geeqie.git;a=commit;h=96a1fc7a86f199045e708f44918aff945bc762e9). The thing is that by the nature of history queue while you are traversing leaves of the tree, back moves you horizontally, not vertically as would be more convenient to me. Thus I implemented an option to change the default behaviour of "back" button to move "up". I'm attaching the patch [geeqie-mouse-back-up.patch.gz](https://github.com/BestImageViewer/geeqie/files/2809964/geeqie-mouse-back-up.patch.gz).

I'm not sure what should happen with "forward" mouse button. Maybe it should be mapped to "back" in this case, but I'm not convinced.

**Comment by tomaszg7** [comment]
Sure, that would be nice, but that requires much more work and will probably introduce many new problems on the way. Note that at the moment, those extra buttons are only processed inside image area, so to achieve this more ambitious work, we would probably need to change quite a bit of code to allow them to register globally.

Hmm, it should also be possible to make the action depend on the area where cursor is placed.

**Comment by caclark** [comment]
Maybe a default mouse handler can be put into layout.c. If so, I don't think it will be too difficult.

Andreas Rönnquist mentioned February 12th as a freeze date for the next Debian release, so I don't think I will commit anything (other than the submitted translation) until after then.

**Comment by tomaszg7** [comment]
Oh, maybe I should work on the translation a bit in this case...

**Comment by gusnan** [comment]
> Maybe a default mouse handler can be put into layout.c. If so, I don't think it will be too difficult.
> 
> Andreas Rönnquist mentioned February 12th as a freeze date for the next Debian release, so I don't think I will commit anything (other than the submitted translation) until after then.

Well, it's getting very tight - I need to package for unstable before that, and the package needs to be in unstable (and without release-critical bugs) for 5 days, and then it gets into testing (which will be the next stable Debian release). So it's getting really tight if it isn't released soon. I'm sorry, I should have specified this more clearly earlier. It _might_ get in after this, but I cannot give any guarantees - It's completely up to the release team, if they find that the changes are non-intrusive enough to warrant it.

**Comment by gusnan** [comment]
And in addition to what I wrote above - I _think_ it might be able to get it in after, since I have already packaged a git snapshot, so there shouldn't be very much changes from this to the released version.

**Comment by mowgli** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Hi,

Am Mi den  6. Feb 2019 um 17:45 schrieb Andreas Rönnquist:
> And in addition to what I wrote above - I _think_ it might be able to get it in after, since I have already packaged a git snapshot, so there shouldn't be very much changes from this to the released version.

I think, it would be a good idea to release one more version before 1.5
or maybe even a version 1.5.

However, I am currently somewhat outside of the development. So I have
to see when I find time to do it.

On the other hand, I have the requirement for myself to produce a best
ever release. :-) That means to fix all that small bugs that got into in
the past.

One of my big problems I have is the GTK+3 integration. For example,
GTK+3 still has troubles with using the mouse wheel for scrolling in a
list. And that is just the bigest usability problem with GTK+3, there
are several smaller ones.

I don't think, that there is a way to solve that in Geeqie than it is
more a GTK+3 bug that is just ignored by upstream. (For example, grisbi
in debian changed recently to GTK+3 and now scrolling in lists is a
pain!)

Regards
   Klaus
- -- 
Klaus Ethgen                                       http://www.ethgen.ch/
pub  4096R/4E20AF1C 2011-05-16            Klaus Ethgen <Klaus@Ethgen.ch>
Fingerprint: 85D4 CA42 952C 949B 1753  62B3 79D0 B06F 4E20 AF1C
-----BEGIN PGP SIGNATURE-----
Comment: Charset: ISO-8859-1

iQGzBAEBCgAdFiEEMWF28vh4/UMJJLQEpnwKsYAZ9qwFAlxbXFwACgkQpnwKsYAZ
9qy9UQv+K2Mdf66qHtF7zN/GydJVxv6lu+qQYl+zRsGzjn4ayE3vvKHXkFpEDBo7
PYHN7Se4LsIXEQJwsjjZub6/i0xI/pQo9Q/7DYLbScmcHuT7JT3kZwuhRMepepNK
YQFo57gyCdNenL8Xo/Z1/O+MiiJKt6duXVC57F/i4LB8oNNBiSehi5klFKRIc3+C
R6mW5+rlsZ55sq6xPcLF7gKmLezKClNesbhL0E2eKDr8LHO/+AHzolvn/ijRBMW8
5jrzS8Du8B58r/HgqxbFVWAkE9Pdz0PYFFL/2K40cvqe1iXbzt5J1pCeI9M9CQ7w
xS2VL9VVDg6y8FFaCMso7tybvZHQAseLGbpDKfa2jRznRJwxTOLle5viNDYisqNv
hzpuvD3LWi0CeaL1oYkEffqXyrvbAJIsd6D9kLO4bQ/4oE5HfNgF8Y8BRSo+KSTp
zB5Hm+BZqvCQDZJIvf6D3Z1oOWqHMCFOodgTDF04t2cAfwTzejBPAh2CFYoPqdZL
1jxsHVnr
=81WG
-----END PGP SIGNATURE-----

**Comment by tomaszg7** [comment]
> One of my big problems I have is the GTK+3 integration.

I just recently observed some weird GTK+3 problem with `tree view` - sometimes I needed double click to enter a folder, and sometimes only a single click. Sometimes Geeqie was also stuck in tree view and I couldn't switch back to list view without restarting.

**Comment by caclark** [comment]
On 6 February 2019 at 22:14:57 -00:00, Klaus Ethgen <notifications@github.com> wrote:

> 
> 
> One of my big problems I have is the GTK+3 integration. For example,
> GTK+3 still has troubles with using the mouse wheel for scrolling in a
> list. And that is just the bigest usability problem with GTK+3, there
> are several smaller ones.
> 
> At some time or other GTK4 will be forced upon the unhappy users.

The GTK3 to GTK4 migration page shows that the "draw" signal has been replaced by something else.
I think that the handling of this signal (which replaced the GTK2 "expose-event" signal) is at the centre of the GTK3 image flicker problem - and possibly the MacOS Mojave problem.

That makes me even more unenthusiastic about trying to solve this problem. I might instead spend some effort to see if the problem still exists with GTK4.


(And I have no problem with mouse wheel scroll)

**Comment by tomaszg7** [comment]
So, what's the final bottom line? Do you want to shelf this issue, until somebody steps in and does a more complete attempt to create a configurable button mapping scheme?

**Comment by caclark** [comment]
> Sometimes Geeqie was also stuck in tree view and I couldn't switch back to list view without restarting.

I noticed that for the first time today. I have opened another bug: #690

> So, what's the final bottom line? Do you want to shelf this issue, until somebody steps in and does a more complete attempt to create a configurable button mapping scheme?

No, I think programmable mouse buttons are the way to go, and I think this is possible to do.
It is just a matter of time.
The recent inclusion of .heic viewing made me think about multi-page viewing for .heic and .tiff files - that is what I have just been looking at. 
If I can't get that done in a reasonable length of time, I will take a look at the mouse problem.

And my garden looks like it could be used as the set for the next Jurassic Park film - don't want any raptors appearing out of the undergrowth :-(

**Comment by caclark** [comment]
Hello Tomasz
The attached .diff is a first attempt at providing user-definable mouse buttons.
If you see any problems with the implementation, please let me know.

The defines are on Edit/Preferences/Behaviour.
[mouse_defines_1.diff.gz](https://github.com/BestImageViewer/geeqie/files/3412016/mouse_defines_1.diff.gz)

**Comment by tomaszg7** [comment]
Buttons 8/9 work fine for me. However I can't get buttons 6/7 to do anything. `xev` recognizes them correctly, I don't know what is the problem. Maybe it's a gtk thing. 

It probably also needs good short names for the buttons to replace the numbers. Or alternatively some kind of button "detector" - i.e. you press a button and the correct number gets highlighted.

I'll keep running the patched version and see if anything else breaks.

**Comment by caclark** [comment]
Hello Tomasz
The mouse I am testing with does not have buttons 6 and 7, so I used xmodmap to switch 6, 7 with 8, 9.
As you pointed out, xev and other programs see the codes, but Geeqie does not see the button events for buttons 6 and 7.
I cannot find any data on this problem, so I decided just to remove the detection of 6 and 7 from the code.
I reverted to the standard naming of buttons 8 and 9.
If you do not find any problems with the attached .diif, I will commit it.
[mouse_defines_2.diff.gz](https://github.com/BestImageViewer/geeqie/files/3413873/mouse_defines_2.diff.gz)

**Comment by tomaszg7** [comment]
The patch seems to work fine.

I checked that previous patch didn't detect buttons 6 and 7 with both Gtk 2 and 3. Moreover I checked how it works in Gimp. Tried current HEAD. They have two relevant controllers for mouse: "Linux input" and "Mouse buttons". The first one provides many triggers including mouse buttons, but I couldn't get it to work for any button (tried 6,7,8,9). The other - lists only mouse buttons, but the list starts with button 8 and goes up to 12. That one works correctly. I don't know if I did something wrong with "Linux input" (or maybe it is broken in development version) but the behaviour of "Mouse buttons" suggest they know that something with 6 and 7 is wrong :)

I was thinking about writing a simple GTK test code to see if these buttons trigger anything, but didn't have time to do it.
